src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1833,6 +1833,7 @@ private TypeParameterConstraintClauseSyntax ParseTypeParameterConstraintClause()
                     while (true)
                     {
                         if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken
+                            || this.CurrentToken.Kind == SyntaxKind.SemicolonToken
```

Comment:
>|| this.CurrentToken.Kind == SyntaxKind.SemicolonToken [](start = 28, length = 54)

It feels like we already should be able to stop at ```;```. For example, we are able to parse ```public abstract void M<T>() where T: class;``` without an error.

Reply:
It looks like this is achieved by taking advantage of ```TerminatorState.IsEndOfMethodSignature```. It feels like it would be more robust to follow the same approach, who knows where else we need to check for the terminator, and it looks like we already have an infrastructure to deal with the task.

---
In reply to: [454654156](https://github.com/dotnet/roslyn/pull/45943#discussion_r454654156) [](ancestors = 454654156)

Reply:
We might be able to drop explicit checks for ```;``` elsewhere if they were added for the benefit of records parsing.

---
In reply to: [454659255](https://github.com/dotnet/roslyn/pull/45943#discussion_r454659255) [](ancestors = 454659255,454654156)

Reply:
1. I made the change to leverage `IsTerminator()` infrastructure in new commit, so that you could see what it looks like, but frankly I think it's not better. Let me know if okay to revert.
If we have an explicit exit here with `{`, then I think it's okay to have an explicit exit for `;` as well. Neither character can be inside a constraint clause.

2. An explicit check for `;` was added for records parsing in `ParseClassOrStructOrInterfaceDeclaration` at line 1477, but I don't think it could benefit from `IsTerminator()` infrastructure.

Reply:
> I made the change to leverage IsTerminator() infrastructure in new commit, so that you could see what it looks like, but frankly I think it's not better. Let me know if okay to revert.
If we have an explicit exit here with {, then I think it's okay to have an explicit exit for ; as well. Neither character can be inside a constraint clause.

I think the approach that relies on ```IsTerminator``` infrastructure is better, assuming we will be able to make changes that I suggested in ```ParseClassOrStructOrInterfaceDeclaration```.

> An explicit check for ; was added for records parsing in ParseClassOrStructOrInterfaceDeclaration at line 1477, but I don't think it could benefit from IsTerminator() infrastructure.

I agree with this, that check is already very specific for records and it makes sense to keep it explicit.

---
In reply to: [454699232](https://github.com/dotnet/roslyn/pull/45943#discussion_r454699232) [](ancestors = 454699232)